src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/mapper/HaFlowMapper.java

```diff
@@ -0,0 +1,79 @@
+/* Copyright 2023 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.flowmonitoring.mapper;
+
+import static java.lang.String.format;
+
+import org.openkilda.model.FlowPath;
+import org.openkilda.model.HaFlow;
+import org.openkilda.model.HaFlowPath;
+import org.openkilda.model.HaSubFlow;
+import org.openkilda.model.PathSegment;
+import org.openkilda.wfm.topology.flowmonitoring.model.FlowState;
+import org.openkilda.wfm.topology.flowmonitoring.model.Link;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Mapper
+public class HaFlowMapper {
+    public static final HaFlowMapper INSTANCE = Mappers.getMapper(HaFlowMapper.class);
+
+    /**
+     * Convert HA-Sub-Flow path representation.
+     */
+    public FlowState toFlowStates(HaSubFlow haSubFlow) {
+        HaFlow haFlow = haSubFlow.getHaFlow();
+
+        if (haFlow != null) {
+            FlowState flowState = new FlowState();
+            flowState.setForwardPath(toLinks(getPathSegments(haFlow.getForwardPath(), haSubFlow.getHaSubFlowId())));
+            flowState.setReversePath(toLinks(getPathSegments(haFlow.getReversePath(), haSubFlow.getHaSubFlowId())));
+            return flowState;
+        }
+        throw new IllegalStateException(format("HA-Flow for HA-Sub-Flow %s is not found.", haSubFlow.getHaFlowId()));
+    }
```

Comment:
We could also use functional approach here like in another methods of this class.


```suggestion
         public FlowState toFlowStates(HaSubFlow haSubFlow) {
        return Optional.ofNullable(haSubFlow.getHaFlow()).map(e -> {
            FlowState flowState = new FlowState();
            flowState.setForwardPath(toLinks(getPathSegments(e.getForwardPath(), haSubFlow.getHaSubFlowId())));
            flowState.setReversePath(toLinks(getPathSegments(e.getReversePath(), haSubFlow.getHaSubFlowId())));
            return flowState;
        }).orElseThrow(() -> new IllegalStateException(format("HA-Flow for HA-Sub-Flow %s is not found.", haSubFlow.getHaFlowId())));
    }
```

Reply:
From my point of view moving to the functional style will not create an additional value but will increase the code complexity and decrease the code readability. So, if you don't mind, I would leave the current solution. 
Thank you for your idea!